stata_kernel/stata_session.py

```diff
@@ -154,261 +136,103 @@ def init_console(self):
-        dot and a space could have pexpect thinking that a result is actually
-        the next prompt. This would be bad and would cause following results to
-        be out of order.
+            self.automate('DoCommandAsync', text)
+            try:
+                self.expect(child=self.automation_log, **kwargs)
+            except KeyboardInterrupt:
+                self.automate('UtilSetStataBreak')
+                self.log_fd.expect('--Break--')
+                self.log_fd.expect('\r?\n\. ')
 
-        Using `\r\n\r\n\. ` is better but I found that the command `shell`
-        returns some ANSI escape codes between lines, and thus I needed to allow
-        for an ANSI escape code.
+        return
 
-        NOTE will need to set timeout to None once sure that running is stable.
-        Otherwise running a task longer than 30s would timeout.
+    def expect(child, md5, kernel=None, display=True):
+        """Watch for end of command from file descriptor or TTY
 
         Args:
-            line (str): literal string ready to send to Stata
-        Returns:
-            (str): unmodified output from line
-            (int): execution time
+            child (pexpect.spawn or fdpexpect.spawn): TTY or log file to watch
+            md5 (str): current value of md5 to watch for
+            kernel (ipykernel.kernelbase): Running kernel to allow me to send back messages from inside this function.
         """
 
-        regex = r'\r\n(\x1b\[\?1h\x1b=)?\r\n\. '
-        self.child.sendline(line)
-        timer = default_timer()
-        try:
-            self.child.expect(regex, timeout=20)
-        except KeyboardInterrupt:
-            self.child.sendcontrol('c')
-            self.child.expect(regex, timeout=20)
-
-        delta = default_timer() - timer
-        return ansi_escape.sub('', self.child.before), delta
-
-    def do_aut_async(self, line):
-        """Run code in Stata Automation using DoCommandAsync
-
-        When running a command with DoCommandAsync, the return code is always 0
-        because the command is just put onto the queue. Stata Automation has the
-        command `UtilStataErrorCode`, but that gives the value of _rc and _rc
-        won't be defined unless you use `cap`. `cap` by default silences output,
-        so I need to use `cap noi`.
-
-        In Stata 15, there is no issue with using multiple of the same prefix,
-        and the last prefix between `qui` and `noi` is the one that determines
-        output showing. So if the user wanted to have it be quiet, it would
-        still be quiet with an extra `cap noi` prefixed.
-
-        Args:
-            line (str): literal string ready to send to Stata
-        Returns:
-            (int): return code from Stata
-            (int): execution time
-        """
+        md5 = '`' + md5 + "'"
+        error_re = r'^r\((\d+)\);'
+        graph_notice = '// This is the stata kernel graph notice'
+        more = r'--more--'
```

Comment:
Shouldn't `more` start with `^`?

I also wonder if there is a way to implement a small (0.1s or something) timeout after matching `more` or `error_re` since a user can print `disp "--more--"` or "disp "r(198);". This wouldn't be an issue with the other stopping conditions, since a user wouldn't be able to print the MD5 hash or the graph notice (at least not without trying).

Reply:
I added `^` to more. I think there are no spaces...

I think those timeouts are something to come back to later if they cause a problem. It's probably pretty rare for someone to want to display `--more--` or `r(198);`.

Reply:
I'm having issues with not being able to stop more actually.

```py
if match_index == 3:
    self.child.send('q')
    break
```
This is sending `q` to the tty whenever it sees `--more--`, but in the TTY it isn't breaking.

![image](https://user-images.githubusercontent.com/15164633/44242660-be068780-a198-11e8-87fc-f079b3f07da7.png)

Reply:
For more it wouldn't be too hard, right? Something like `expect(r"^(?!--more--\Z).+", timeout = 0.1)` and if it times out then there is nothing printed after `--more--`. The error would be trickier. Not sure what is a good idea for that.

Reply:
You need to send about 3 newlines/carriage returns because when you quit more it wants to give you `error 1`

Reply:
Because then `--more--` wouldn't be caught by that... It would be caught by one of the other rules, like the end of line. Or do you mean I'd do that after I've already successfully expected for `^--more--`?

Reply:
Yeah. If you match more, then expect something that isn't more. If you don't get it then `--more--` was a stopping point. Otherwise it was printed by the user for some reason. That regex should match anything that isn't `^--more--$`

Reply:
Fun fact: When I re-size my terminal, the pagesize resets to whatever Stata now thinks is reasonable. I doubt this will ever be caught out by that (though sometimes it seems you need more than 3 CR to break `more`...

Reply:
> You need to send about 3 newlines/carriage returns because when you quit more it wants to give you error 1

In my terminal, when I press `q`, the more breaks. Then I need to expect for 3 newlines, maybe, but I don't need to send three newlines. Here, sending q isn't doing anything.

Reply:
Note to self, apparently `ctrl-C` then `ctrl-D` also breaks a `more`. Test that out with expect.